zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
+	[ INFO] [1635948213.689598629]: Calculated depth is 113.001 ✅
+	[ INFO] [1635948229.253951900]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_goal_behind_power_cell.png
+	[ INFO] [1635948229.870174338]: Calculated depth is 138.001 ✅
+	[ INFO] [1635948236.848476178]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_goal.png
+	[ INFO] [1635948237.445102845]: Calculated depth is 113.001 ✅
+	[ INFO] [1635948248.428153064]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_loading_bay.png
+	[ INFO] [1635948249.227361594]: Calculated depth is 141.001 ✅
+	[ INFO] [1635948255.684437760]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_low_goal.png
+	[ INFO] [1635948256.286312490]: Calculated depth is 92.001 ✅
+	[ INFO] [1635948266.234706198]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_power_cell.png
+	[ INFO] [1635948266.316279430]: Calculated depth is 0.001 ✅
+	[ INFO] [1635948273.702984310]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/goal_behind_power_cell.png
+	[ INFO] [1635948275.086796499]: Calculated depth is 0.001 **non-cropped version** ❌
+	[ INFO] [1635948302.568242461]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/goal_with_power_cell_in_front.jpg
+	[ INFO] [1635948302.598558216]: Calculated depth is 0.001 ❌
+	[ INFO] [1635948312.950812216]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/test_weirdness.jpg
+	[ INFO] [1635948312.980837687]: Calculated depth is 73.001 ✅
+	*/
+}
+
+// Get the most useful depth value in the cv::Mat depth contained within
+// the supplied bounding rectangle, using k-means
+double kMeansDepthMat(const cv::Mat& depth, const cv::Rect& bound_rect, bool debug = false, int maximumK = 3, float tolerance = 1e-3)
 {
-	double sum = 0.0;
-	unsigned count = 0;
-	for (int j = bound_rect.tl().y+1; j < bound_rect.br().y; j++) //for each row
-	{
-		const float *ptr_depth = depth.ptr<float>(j);
+	// setup randomizing (for initialization of k-means)
+	std::random_device seeder;
+	std::mt19937 engine(seeder());
+
+	// Start of checking if k is too high
 
-		for (int i = bound_rect.tl().x+1; i < bound_rect.br().x; i++) //for each pixel in row
+	// Convert cropped depth matrix to vector
+	std::vector<float> vec;
+	cv::Mat mat = depth(bound_rect);
+	for (int i = 0; i < mat.rows; i++) {
+    vec.insert(vec.end(), mat.ptr<float>(i), mat.ptr<float>(i)+mat.cols*mat.channels());
+  }
+
+	// Find unique (defined as >1cm away) values in the vector
+	std::sort(vec.begin(), vec.end());
+	auto last = std::unique(vec.begin(), vec.end(), [](float first, float second){ return fabs(second - first)<0.01; });
```

Comment:
Here's a chance to leave a "no magic numbers" comment.  Here, 0.01 feels like it could be a config value to tweak once we get more data, and putting it somewhere easy to find would make that easier.  If you haven't seen it, c++ lets you define constants using

`constexpr double depth_epsilon = 0.01;`

Put this near the top of the file, and later on when we talk about params it'll be easy to un-constexpr the value and hook it up as a configurable parameter read at startup.  And even better, using http://wiki.ros.org/dynamic_reconfigure/Tutorials/SettingUpDynamicReconfigureForANode%28cpp%29 it could also be set to be changeable in real-time for testing.  Don't use that exact example, it'll be easy to  do with our dynamic reconfigure code (e.g. https://github.com/FRC900/ddynamic_reconfigure) instead.

Reply:
I added the `constexpr`, and the dynamic reconfiguration sounds cool -- I'm happy to implement that once we can test it